### PR TITLE
im-util: Re-introduce ArrayVec for limit adapter buffering

### PR DESCRIPTION
**Pull request by jplatte**

### CODE REVIEW

File: eyeball-im-util/src/vector.rs [hunk 1]
```diff
@@ -41,6 +41,11 @@ type VectorDiffContainerStreamFamily<S> =
...
-type VectorDiffContainerStreamBuffer<S> =
-    <<S as Stream>::Item as VectorDiffContainerOps<VectorDiffContainerStreamElement<S>>>::Buffer;
+/// [`VectorDiffContainer`]s' `LimitBuf`.
+type VectorDiffContainerStreamLimitBuf<S> =
+    <<S as Stream>::Item as VectorDiffContainerOps<VectorDiffContainerStreamElement<S>>>::LimitBuf;
+
+/// Type alias for extracting the buffer type from a stream of
+/// [`VectorDiffContainer`]s' `SortBuf`.
+type VectorDiffContainerStreamSortBuf<S> =
+    <<S as Stream>::Item as VectorDiffContainerOps<VectorDiffContainerStreamElement<S>>>::SortBuf;
```
Comment (Hywan) [inline eyeball-im-util/src/vector.rs:51]:
```suggestion
/// [`VectorDiffContainer`]s' `ArrayBuffer`.
type VectorDiffContainerStreamArrayBuffer<S> =
    <<S as Stream>::Item as VectorDiffContainerOps<VectorDiffContainerStreamElement<S>>>::ArrayBuffer;

/// Type alias for extracting the buffer type from a stream of
/// [`VectorDiffContainer`]s' `VecBuffer`.
type VectorDiffContainerStreamVecBuffer<S> =
    <<S as Stream>::Item as VectorDiffContainerOps<VectorDiffContainerStreamElement<S>>>::VecBuffer;
```

File: eyeball-im-util/src/vector/ops.rs [hunk 2]
```diff
@@ -1,9 +1,11 @@
+use arrayvec::ArrayVec;
 use eyeball_im::VectorDiff;
 use smallvec::SmallVec;
 
 pub trait VectorDiffContainerOps<T>: Sized {
     type Family: VectorDiffContainerFamily;
-    type Buffer: Default;
+    type LimitBuf: Default;
+    type SortBuf: Default;
```
Comment (Hywan) [inline eyeball-im-util/src/vector/ops.rs:8]:
```suggestion
    type ArrayBuffer: Default;
    type VecBuffer: Default;
```

File: eyeball-im-util/src/vector/ops.rs [hunk 3]
```diff
@@ -12,21 +14,30 @@ pub trait VectorDiffContainerOps<T>: Sized {
...
 
-    fn push_into_buffer(
+    fn push_into_limit_buf(
         self,
-        buffer: &mut Self::Buffer,
+        buffer: &mut Self::LimitBuf,
+        make_diffs: impl FnMut(VectorDiff<T>) -> ArrayVec<VectorDiff<T>, 2>,
+    ) -> Option<Self>;
+
+    fn pop_from_limit_buf(buffer: &mut Self::LimitBuf) -> Option<Self>;
```
Comment (Hywan) [inline eyeball-im-util/src/vector/ops.rs:23]:
```suggestion
    fn pop_from_array_buffer(buffer: &mut Self::ArrayBuffer) -> Option<Self>;
```

File: eyeball-im-util/src/vector/ops.rs [hunk 4]
```diff
@@ -12,21 +14,30 @@ pub trait VectorDiffContainerOps<T>: Sized {
...
+
+    fn push_into_sort_buf(
```
Comment (Hywan) [inline eyeball-im-util/src/vector/ops.rs:25]:
```suggestion
    fn push_into_vec_buffer(
```

File: eyeball-im-util/src/vector/ops.rs [hunk 5]
```diff
@@ -12,21 +14,30 @@ pub trait VectorDiffContainerOps<T>: Sized {
...
+        self,
+        buffer: &mut Self::SortBuf,
         make_diffs: impl FnMut(VectorDiff<T>) -> SmallVec<[VectorDiff<T>; 2]>,
     ) -> Option<Self>;
 
-    fn pop_from_buffer(buffer: &mut Self::Buffer) -> Option<Self>;
+    fn pop_from_sort_buf(buffer: &mut Self::SortBuf) -> Option<Self>;
```
Comment (Hywan) [inline eyeball-im-util/src/vector/ops.rs:31]:
```suggestion
    fn pop_from_vec_buffer(buffer: &mut Self::VecBuffer) -> Option<Self>;
```

File: eyeball-im-util/src/vector/ops.rs [hunk 6]
```diff
@@ -12,21 +14,30 @@ pub trait VectorDiffContainerOps<T>: Sized {
         f: impl FnMut(VectorDiff<T>) -> Option<VectorDiff<U>>,
     ) -> Option<VectorDiffContainerFamilyMember<Self::Family, U>>;
 
-    fn push_into_buffer(
+    fn push_into_limit_buf(
```
Comment (Hywan) [inline eyeball-im-util/src/vector/ops.rs:17]:
```suggestion
    fn push_into_array_buffer(
```

**Review by Hywan — COMMENTED** [review]
I am not a fan of this, but I can live with it.

Can we please rename `LimitBuf` to `ArrayBuffer` and `SortBuf` to `VecBuffer`? It becomes generic and can be re-used by other stream adapters.
If we go that path, we can also renaùe `VectorDiffContainerStreamLimitBuf` to `VectorDiffContainerStreamArrayBuffer` and `VectorDiffContainerStreamSortBuf` to `VectorDiffContainerStreamVecBuffer`.

I've made a couple of suggestions during the review.

**Comment by jplatte** [comment]
> I am not a fan of this, but I can live with it.

What do you not like? The bit of code duplication in `ops.rs`? Having two dependencies for similar tasks?

> Can we please rename `LimitBuf` to `ArrayBuffer` and `SortBuf` to `VecBuffer`? It becomes generic and can be re-used by other stream adapters.

I think something like that should only be done once another adapter is actually written that uses the same types. Note that `SortBuf` is already more general than the current `SortBy` adapter, because I think we'll add `Sort` and maybe also `SortByKey` soon and they'll clearly need the exact same buffering logic.

**Comment by Hywan** [comment]
> What do you not like? The bit of code duplication in ops.rs? Having two dependencies for similar tasks?

Yes, the code “duplication” and the two dependencies. I understand why you prefer a static array: it statically asserts that we won't “overflow”/overgrow the buffer. I wish `SmallVec` could have an inline-only type (in this case, it would strictly be like `ArrayVec`). I just believe that having a single type would simplify the API of `VectorDiffContainerOps`. Right now, it seems confusing for a first sight. It's also possible to use both at the same time in the same `Stream`, which sounds a bit confusing too. What do you think?

It's OK to not agree. In this case, the patch is fine and I'm OK to approve it too.

**Comment by jplatte** [comment]
Time for `ArrayVec` to be added to the standard library x)

There is actually the `tinyvec` crate as an alternative that provides both an `ArrayVec` and `TinyVec` (like `SmallVec`), but unfortunately it requires the inner type to implement `Default`.

Do you think it would make sense to feature-gate the individual adapters (or group of adapters like `Filter` + `FilterMap`, `Sort` + `SortBy` + `SortByKey`) in the next breaking-change release? It wouldn't make a difference for the Matrix Rust SDK, but this PR doesn't make a difference for it either (I found both `arrayvec` and `smallvec` are already in the dependency graph anyways).

**Comment by Hywan** [comment]
I don't believe it makes sense to feature-flag them for the moment. We don't have a lot of them.

**Pull request merged**